src/validation.cpp

```diff
@@ -3010,13 +3010,43 @@ bool CChainState::ActivateBestChainStep(CValidationState& state, const CChainPar
                     state = CValidationState();
                     if (pindexConnect == pindexMostWork) {
-                        // NOTE: Invalidate blocks back to last checkpoint
+                        //checkpoints
                         auto &checkpoints = chainparams.Checkpoints().mapCheckpoints;
-                        auto it = checkpoints.lower_bound(pindexConnect->nHeight);
-                        if (it != checkpoints.begin() && (--it)->first > 0) { // it doesn't makes sense backward to genesis
-                            auto index = LookupBlockIndex(it->second);
-                            if (!disconnectBlocksTo(index)) {
-                                return false;
+                        //calculate the latest suitable checkpoint block height
+                        auto checkpointIt = checkpoints.lower_bound(pindexConnect->nHeight);
+                        auto fallbackCheckpointBlockHeight = (checkpointIt != checkpoints.begin()) ? (--checkpointIt)->first : 0;
+
+                        auto invalidateBlocksTo = [this, &disconnectBlocksTo](const uint256& hash) -> bool {
+                            auto index = LookupBlockIndex(hash);
+                            return disconnectBlocksTo(index);
+                        };
+
+                        //check spv and anchors not available
+                        if (!spv::pspv || !panchors) {
+                            // Invalidate blocks back to fallbackCheckpointBlockHeight
+                            if (fallbackCheckpointBlockHeight > 0) { // it doesn't makes sense backward to genesis
+                                if (!invalidateBlocksTo(checkpointIt->second)) {
+                                    return false;
+                                }
+                            }
+                        } else {
+                            //get most latest anchor which is lower than  pindexConnect->nHeight
+                            auto fallbackAnchor = panchors->GetMostLatestAtDeFiHeight(pindexConnect->nHeight);
+
+                            if (fallbackAnchor && (fallbackAnchor->anchor.height >= fallbackCheckpointBlockHeight))
+                            {
+                                // Invalidate blocks back to fallbackkAnchor height
+                                if (!invalidateBlocksTo(fallbackAnchor->anchor.blockHash)) {
+                                    return false;
+                                }
+                            }
+                            else {
+                                //No latest fallbackAnchor, Invalidate blocks back to fallbackCheckpointBlockHeight
+                                if (fallbackCheckpointBlockHeight > 0) { // it doesn't makes sense backward to genesis
+                                    if (!invalidateBlocksTo(checkpointIt->second)) {
+                                        return false;
+                                    }
+                                }
```

Comment:
It looks over-engineering, just make an `uint256 checkpointHash` get it and disconnect blocks at end.

Reply:
That's the old implementation. The PR is to use anchor height if it is available, if not back to checkpointHash

Reply:
No, new code is over-engineering. Which is one is more clear
```
CBlockIndex *blockIndex = nullptr;
if (spv::pspv && panchors) { // try anchors first
    auto fallbackAnchor = panchors->GetMostLatestAtDeFiHeight(pindexConnect->nHeight);
    if (fallbackAnchor && (fallbackAnchor->anchor.height >= fallbackCheckpointBlockHeight)) {
        blockIndex = LookupBlockIndex(fallbackAnchor->anchor.blockHash);
    }
}
if (!blockIndex && fallbackCheckpointBlockHeight > 0) { // fallback
    blockIndex = LookupBlockIndex(checkpointIt->second);
}
if (blockIndex) {
    disconnectBlocksTo(blockIndex);
}
```

Reply:
lesser LOC agreed. but functionality wise it's the same.
will update with a minor change as the following code. thanks
```
CBlockIndex *blockIndex = nullptr;
if (spv::pspv && panchors) { // try anchors first
    auto fallbackAnchor = panchors->GetMostLatestAtDeFiHeight(pindexConnect->nHeight);
    if (fallbackAnchor && (fallbackAnchor->anchor.height > fallbackCheckpointBlockHeight)) {
        blockIndex = LookupBlockIndex(fallbackAnchor->anchor.blockHash);
    }
}
if (!blockIndex && fallbackCheckpointBlockHeight > 0) { // fallback
    blockIndex = LookupBlockIndex(checkpointIt->second);
}
if (blockIndex) {
    disconnectBlocksTo(blockIndex);
}
```